book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,691 @@
+| `block_by_height`    | `BE32(height)`        | `Block`                             |
+| `tx_by_hash`         | `transaction::Hash`   | `BE32(height) || BE32(tx_index)`    |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+| `sprout_anchors`     | `sprout::tree::Root`  | `()`                                |
+| `sapling_anchors`    | `sapling::tree::Root` | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
+### Notes on Sled trees
+
+- The `hash_by_height` and `height_by_hash` trees provide the bijection between
+  block heights and block hashes.  (Since the Sled state only stores finalized
+  state, this is actually a bijection).
+
+- Blocks are stored by height, not by hash.  This has the downside that looking
+  up a block by hash requires an extra level of indirection.  The upside is
+  that blocks with adjacent heights are adjacent in the database, and many
+  common access patterns, such as helping a client sync the chain or doing
+  analysis, access blocks in (potentially sparse) height order.  In addition,
+  the fact that we commit blocks in order means we're writing only to the end
+  of the Sled tree, which may help save space.
+
+- Transaction references are stored as a `(height, index)` pair referencing the
+  height of the transaction's parent block and the transaction's index in that
+  block.  This would more traditionally be a `(hash, index)` pair, but because
+  we store blocks by height, storing the height saves one level of indirection.
+
+## Committing finalized blocks
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.  Otherwise, commit the block described below, then
+commit any queued children.  (Although the checkpointer generates verified
+blocks in order when it completes a checkpoint, the blocks are committed in the
+response futures, so they may arrive out of order).
+
+Committing a block to the sled state should be implemented as a wrapper around
+a function also called by [`Request::CommitBlock`](#request-commit-block),
+which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic. This check is performed as defense-in-depth to
+prevent database corruption, but it is the caller's responsibility (e.g. the
+zebra-state service's responsibility) to commit finalized blocks in order.
+
+2. Insert:
+    - `(hash, height)` into `height_by_hash`;
+    - `(height, hash)` into `hash_by_height`;
+    - `(height, block)` into `block_by_height`.
+
+3.  Update the `sprout_anchors` and `sapling_anchors` trees with the Sprout
+    and Sapling anchors (XXX: how??)
+
```

Comment:
```suggestion
3.  Update the `sprout_anchors` and `sapling_anchors` trees with the Sprout
    and Sapling anchors. The Sprout and Sapling anchors are the roots of the Sprout and Sapling note commitment trees that have already been calculated for the last transaction(s) in the block that have `JoinSplit`s in the Sprout case and/or `Spend`/`Output` descriptions in the Sapling case.

```

Reply:
@hdevalence @yaahc does this make sense?

Reply:
Suggest "sled trees". It was confusing to me that there were extra trees in the sled db before it was explained.

Reply:
> does this make sense?

Yea, the sentence runs on a bit longer than I'd like but I can't for the life of me figure out how I'd break it up. That said I think it might be good to move the explanation of how out of the bullet and into a note after like this note https://github.com/ZcashFoundation/zebra/pull/902/files#diff-0a5b73420cea8c2ce73f46bc2b5d56c7R92.

Reply:
@yaahc updated. Is that the correct location for the **note** or should it go under all the bullets?

Reply:
I lean towards moving it under the bullets but im merging this and ill move it manually as part of my next edit pass